### PR TITLE
test: extend the SOCKS5 Python proxy to actually connect to a destination

### PR DESCRIPTION
**Pull request by vasild**
If requested, make the SOCKS5 Python proxy redirect connections to a set of given destinations. Actually act as a real proxy, connecting the client to a destination, except that the destination is not what the client asked for.

This would enable us to "connect" to Tor addresses from the functional tests.

Plus a few other minor improvements in the test framework as individual commits.

---

These changes are part of https://github.com/bitcoin/bitcoin/pull/29415 but they make sense on their own and would be good to have them, regardless of the fate of #29415. Also, if this is merged, that would reduce the size of #29415, thus the current standalone PR.

### CODE REVIEW

File: test/functional/test_framework/p2p.py [hunk 1]
```diff
@@ -188,6 +188,10 @@ def peer_connect_helper(self, dstaddr, dstport, net, timeout_factor):
         self.on_connection_send_msg = None
         self.recvbuf = b""
         self.magic_bytes = MAGIC_BYTES[net]
+        if dstport != 0:
+            self.p2p_connected_to_node = True
+        else:
+            self.p2p_connected_to_node = False
```
Comment (Empact) [inline test/functional/test_framework/p2p.py:194]: nit: you can assign the comparison directly

**Comment by vasild** [comment]
`05f8c032b8...4267c206f4`: rebase due to conflicts and address https://github.com/bitcoin/bitcoin/pull/29420#discussion_r1503740908

File: test/functional/test_framework/p2p.py (hunk 1 above)
Comment (vasild) [reply test/functional/test_framework/p2p.py]: Done

**Comment by vasild** [comment]
`4267c206f4...9c50b2fe44`: pick changes due to suggestions from https://github.com/bitcoin/bitcoin/pull/29415

**Review by pinheadmz — APPROVED** [review]
ACK 9c50b2fe44e4744204c51b82e86174db1d749cfe

Reviewed each commit and left some questions below. Built and ran all unit and functional tests on arm/macos. I am already familiar with this feature from #29415 and have run those tests as well.

One nitty nitty nit nit: commit message 9a9b545c7fc044242c4dd8a4916b2ba13b3ea850 uses c++ syntax instead of python ;-)


<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 9c50b2fe44e4744204c51b82e86174db1d749cfe
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmYgKwwACgkQ5+KYS2KJ
yTptJQ/+OksRFOpcFGtbEYIaaAaiWAlLJVBQYa2ljpRvSxecoWqmcgP6wvaXGSsA
SxBiu7Y+1fZqBEaRLeOLm7GtZiZesc87Cu2BJGM5SvxLkiF6l7YqtwISq51Z0cZu
C15s/eSexvpLFp+tGvlcm9au5iZShU7nGMVNAEz5vfFUJOQBDVDJ1mib1hxaLRaV
4U97uoYQcOVZDz+fdzisekP1wpklXZTDQiS9grbHMBmqeT45ugxgUEC+DVWFtm5W
01JBblBPNLvlOFiNpHCy0MUNXtg7n1ER2ee677ZTP05ck9hOga0gu8g8doemvZ91
QPs2nLq0dHvaFtlYl2N3Ce3aRBX6I3qfegKWmrBWm54KB3mOSrV3OP2j0UMUGmcM
nwPQruUg+GnBsEWK9RgvLAmSFbkGEJzS0nqFfOkRLLEwNWAzMUYAooXs4J7xLVcA
AR8ncBnz9zVBkdhrHbT8wYcf22N/W+k3fwxCgTxHPiAt10UXAzkdfnRfwjjQG7NU
xmP20eAFsE9FWOnfkVIh/VCzcgS+wT629gaVYJqF9hX16K9uKH2AuOQeCLnRiX3D
E1B1YnspQklZAy0ejc2BZEuwBc+MjtSEowTPK9ENT8NJoEO2y3YrnGcbXThyVrBQ
yvLU1dfljFfSq2seaNP4fpKJT8cKXAdCHj/HlowVsYw9hDs/PKU=
=VgCO
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

File: test/functional/test_framework/p2p.py [hunk 2]
```diff
@@ -217,7 +217,12 @@ def peer_disconnect(self):
     def connection_made(self, transport):
         """asyncio callback when a connection is opened."""
         assert not self._transport
-        logger.debug("Connected & Listening: %s:%d" % (self.dstaddr, self.dstport))
+        info = transport.get_extra_info("socket")
+        us = info.getsockname()
+        them = info.getpeername()
+        logger.debug(f"Connected: us={us[0]}:{us[1]}, them={them[0]}:{them[1]}")
+        self.dstaddr = them[0]
+        self.dstport = them[1]
```
[the remark this answers is not in the document]
Comment (vasild) [reply test/functional/test_framework/p2p.py:226]: I set `dstaddr` and `dstport` here in `connection_made()` because when accepting connections (from the python node point of view) they are set to `"0"` and `0` respectively (in `master`):

https://github.com/bitcoin/bitcoin/blob/ef09f535b7b540d997f8c2bfa67b3386711bc8b4/test/functional/test_framework/p2p.py#L203

https://github.com/bitcoin/bitcoin/blob/ef09f535b7b540d997f8c2bfa67b3386711bc8b4/test/functional/test_framework/p2p.py#L182-L186

When I add those asserts I get:
```
    assert self.dstaddr == them[0]
           ^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

File: test/functional/test_framework/socks5.py [hunk 3]
```diff
@@ -117,6 +128,33 @@ def handle(self):
...
+                    self.conn.setblocking(False)
+                    conn_to.setblocking(False)
+                    sockets = [self.conn, conn_to]
+                    done = False
+                    while not done:
+                        rlist, _, xlist = select.select(sockets, [], sockets)
+                        if len(xlist) > 0:
+                            raise IOError("Exceptional condition on socket")
+                        for s in rlist:
+                            data = s.recv(4096)
```
[the remark this answers is not in the document]
Comment (vasild) [reply test/functional/test_framework/socks5.py:147]: No, I just picked some value that looks reasonable, anything would work. For example `1` or `5` would still work but does not look reasonable because it will cause a lot of iteration loops.

File: test/functional/test_framework/socks5.py [hunk 4]
```diff
@@ -41,6 +42,16 @@ def __init__(self):
...
+        # An array of objects like:
+        # {
+        #     "to_addr": "127.0.0.1"
+        #     "to_port": 28276
+        #     "node": python_p2p_node,
+        #     "requested_to_addr": "the_client_asked_to_connect_to_this_addr.onion",
+        # }
+        # Redirect the i'th connection to destinations[i] to_addr:to_port
+        self.destinations = []
+        self.destinations_used = 0
```
[the remark this answers is not in the document]
Comment (vasild) [reply test/functional/test_framework/socks5.py:54]: I slightly prefer to keep the array unmodified in case the data in it is needed at some later point (for logging or something).

File: test/functional/test_framework/p2p.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vasild) [reply test/functional/test_framework/p2p.py:226]: from comment https://github.com/bitcoin/bitcoin/pull/29415#discussion_r1659087156:
> Why didn't we need to do this on master?
> The old log message was `Connected & Listening: 127.0.0.1:15689` not `Connected & Listening: 0:0`

When accepting connections (from the python node point of view), then indeed 0:0 is printed (on `master`):

```
$ ./test/functional/p2p_addr_relay.py -l debug --nocleanup
$ ./test/functional/combine_logs.py |grep 'Connected & L'
...
 test  2024-07-02T10:18:18.494000Z TestFramework.p2p (DEBUG): Connected & Listening: 127.0.0.1:14829 
 test  2024-07-02T10:18:18.609000Z TestFramework.p2p (DEBUG): Connected & Listening: 127.0.0.1:14829 
 test  2024-07-02T10:18:20.556000Z TestFramework.p2p (DEBUG): Connected & Listening: 0:0 
 test  2024-07-02T10:18:20.719000Z TestFramework.p2p (DEBUG): Connected & Listening: 0:0 
 test  2024-07-02T10:18:20.934000Z TestFramework.p2p (DEBUG): Connected & Listening: 127.0.0.1:14829 
 test  2024-07-02T10:18:22.265000Z TestFramework.p2p (DEBUG): Connected & Listening: 0:0 
```

**Comment by vasild** [comment]
`9c50b2fe44...a179b1cbf5`: rebase due to conflicts

**Comment by vasild** [comment]
`a179b1cbf5...5956668ecd`: rebase and pick changes from https://github.com/bitcoin/bitcoin/pull/29415#issuecomment-2260523021, more specifically:

> Change the extension of the SOCKS5 proxy - instead of being given a set of predefined destinations to redirect the connections to, it is now given a callback function which returns the destination where to redirect, or it can decide to return None asking the SOCKS5 proxy to close the connection (default behavior). This is more flexible.

**Comment by vasild** [comment]
`5956668ecd...f93fab4c58`: rebase to pick CMake (even though this PR only contains changes to `.py` files)

**Comment by achow101** [comment]
cc @tdb3

File: test/functional/test_framework/socks5.py [hunk 5]
```diff
@@ -117,6 +135,38 @@ def handle(self):
...
+                dest = self.serv.conf.destinations_factory(requested_to_addr, port)
+                if dest is not None:
+                    logger.debug(f"Serving connection to {requested_to}, will redirect it to "
+                                 f"{dest['actual_to_addr']}:{dest['actual_to_port']} instead")
+                    with socket.create_connection((dest["actual_to_addr"], dest["actual_to_port"])) as conn_to:
+                        self.conn.setblocking(False)
+                        conn_to.setblocking(False)
+                        sockets = [self.conn, conn_to]
+                        done = False
+                        while not done:
```
Comment (laanwj) [inline test/functional/test_framework/socks5.py:152]: Maybe factor this shovel-from-socketa-to-socketb loop out to a function in netutil as well.

File: test/functional/test_framework/socks5.py [hunk 6]
```diff
@@ -117,6 +135,38 @@ def handle(self):
...
+                            rlist, _, xlist = select.select(sockets, [], sockets)
+                            if len(xlist) > 0:
+                                raise IOError("Exceptional condition on socket")
+                            for s in rlist:
+                                data = s.recv(4096)
+                                if data is None or len(data) == 0:
+                                    done = True
+                                    break
+                                if s == self.conn:
+                                    conn_to.sendall(data)
```
Comment (laanwj) [inline test/functional/test_framework/socks5.py:162]: i don't think `sendall` can be guaranteed with non-blocking sockets, it will fail if it gets EAGAIN
https://stackoverflow.com/questions/6240737/python-socket-sendall-function

**Comment by vasild** [comment]
`f93fab4c58...57529ac4db`: address suggestions

File: test/functional/test_framework/socks5.py (hunk 5 above)
Comment (vasild) [reply test/functional/test_framework/socks5.py]: I moved it to its own function, but in `test/functional/test_framework/socks5.py` because it uses the newly created `sendall()` (due to the [suggestion below](https://github.com/bitcoin/bitcoin/pull/29420#discussion_r1802683914)) which is very similar to `recvall()` which already exists in `socks5.py`.

I guess if this is to be moved to `test/functional/test_framework/netutil.py`, then all 3 of `recvall()`, `sendall()` and `forward_sockets()` should be moved. Leave it as is or move to `netutil.py`?

File: test/functional/test_framework/socks5.py (hunk 6 above)
Comment (vasild) [reply test/functional/test_framework/socks5.py]: Ok, [`sendall()`](https://docs.python.org/3/library/socket.html#socket.socket.sendall) doc says "On error, an exception is raised, and there is no way to determine how much data, if any, was successfully sent" so this is useless for us. The doc says as well that since Python 3.5 the method will retry if the syscall is interrupted, but mentions nothing about `EAGAIN` or `EWOULDBLOCK`.

So I followed https://docs.python.org/3/howto/sockets.html#using-a-socket and used the basic `send()` with a manual retry loop, like in that example.

Diving further into Python source code, [`sendall()`](https://github.com/python/cpython/blob/b454662921fd3a1fc27169e91aca03aadea08817/Modules/socketmodule.c#L4418) [calls](https://github.com/python/cpython/blob/b454662921fd3a1fc27169e91aca03aadea08817/Modules/socketmodule.c#L4461) [ `sock_call_ex()`](https://github.com/python/cpython/blob/b454662921fd3a1fc27169e91aca03aadea08817/Modules/socketmodule.c#L886) which is [retrying](https://github.com/python/cpython/blob/b454662921fd3a1fc27169e91aca03aadea08817/Modules/socketmodule.c#L994-L1004) on `EWOULDBLOCK` and `EAGAIN`. This is not documented explicitly anywhere, so I decided to ignore it and go with the manual retry loop. Also, `send()` (in addition to `sendall()`) ends up with that retry, so maybe the manual retry loop will always loop just once.

File: test/functional/test_framework/socks5.py [hunk 7]
```diff
@@ -117,6 +171,22 @@ def handle(self):
...
+            requested_to_addr = addr.decode("utf-8")
+            requested_to = format_addr_port(requested_to_addr, port)
+
+            if self.serv.conf.destinations_factory is not None:
+                dest = self.serv.conf.destinations_factory(requested_to_addr, port)
+                if dest is not None:
+                    logger.debug(f"Serving connection to {requested_to}, will redirect it to "
+                                 f"{dest['actual_to_addr']}:{dest['actual_to_port']} instead")
+                    with socket.create_connection((dest["actual_to_addr"], dest["actual_to_port"])) as conn_to:
+                        forward_sockets(self.conn, conn_to)
```
Comment (vasild) [inline test/functional/test_framework/socks5.py:184]: Following discussion with @tdb3, how do we know this code even works and is not totally broken? That is a reasonable question because this PR does not demonstrate that this works.

https://github.com/bitcoin/bitcoin/pull/29415 (which includes this PR) exercises the newly added functionality. The tests from https://github.com/bitcoin/bitcoin/pull/29415 that use this depend on the private broadcast functionality so can't be included here.

**Review by tdb3 — APPROVED** [review]
CR and test ACK 57529ac4dbb2721c1ad0a3566f0299dbdb5ca5c0

Checked with a hacky little sanity test where a connection was made to the proxy on 127.0.0.1:10000 (set up to detour to 127.0.0.1:10001).  A small echo server listened on 10001 (echoing back "testack").  The new detouring functionality of the proxy seemed to work as expected.

<details>
<summary>hacky test</summary>

```
build/test/functional/feature_framework_unit_tests.py
```

```diff
diff --git a/test/functional/feature_framework_unit_tests.py b/test/functional/feature_framework_unit_tests.py
index 14d83f8a70..9bbfcf987e 100755
--- a/test/functional/feature_framework_unit_tests.py
+++ b/test/functional/feature_framework_unit_tests.py
@@ -29,6 +29,7 @@ TEST_FRAMEWORK_MODULES = [
     "script",
     "script_util",
     "segwit_addr",
+    "socks5",
     "wallet_util",
 ]
 
diff --git a/test/functional/test_framework/socks5.py b/test/functional/test_framework/socks5.py
index 3387c8a1fe..e18d62ee3c 100644
--- a/test/functional/test_framework/socks5.py
+++ b/test/functional/test_framework/socks5.py
@@ -7,6 +7,7 @@
 import select
 import socket
 import threading
+import unittest
 import queue
 import logging
 
@@ -230,3 +231,28 @@ class Socks5Server():
         s.connect(self.conf.addr)
         s.close()
         self.thread.join()
+
+class TestFrameworkSocks5(unittest.TestCase):
+    def test_destinations(self):
+        socks_conf = Socks5Configuration()
+        socks_conf.addr = ('127.0.0.1', 10000)
+        socks_conf.unauth = True
+        socks_conf.auth = False
+        socks_conf.destinations_factory = lambda addr, port : {"actual_to_addr": '127.0.0.1', "actual_to_port": 10001}
+        socks_serv = Socks5Server(socks_conf)
+        socks_serv.start()
+        # send initial greeting, receive response
+        client_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        client_sock.connect(('127.0.0.1', 10000))
+        sendall(client_sock, bytearray([0x05, 0x01, 0x00]))
+        response = recvall(client_sock, 2)
+        assert response == bytearray([0x05, 0x00])
+        # send connect (to 127.0.0.1:10000), receive response
+        sendall(client_sock, bytearray([0x05, 0x01, 0x00, 0x03, 0x09, 0x31, 0x32, 0x37, 0x2e, 0x30, 0x2e, 0x30, 0x2e, 0x31, 0x27, 0x10]))
+        response = recvall(client_sock, 10)
+        assert response == bytearray([0x05, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
+        # send "test" and observe response on 10001 (e.g. with echo server)
+        sendall(client_sock, b'test')
+        response = recvall(client_sock, 7)
+        assert response == b'testack'
+        socks_serv.stop()
```
</details>

<details>
<summary>echo server</summary>

```python
import socket

HOST = "127.0.0.1"
PORT = 10001

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.bind((HOST, PORT))
    s.listen()
    conn, addr = s.accept()
    with conn:
        while True:
            data = conn.recv(1024)
            if not data:
                break
            conn.sendall(data + b'ack')
```
</details>

**Review by mzumsande — COMMENTED** [review]
Code review / tested ACK 57529ac4dbb2721c1ad0a3566f0299dbdb5ca5c0

This is pretty cool, and could be used in other places than just for the private transaction relay (e.g. whenever we'd like to have an onion outbound peer in a functional test and exchange messages with it)

I tested this by extending `feature_anchors.py` to use a `destinations_factory` and checked that it completes the version handshake and behaves like a normal connection.

File: test/functional/test_framework/socks5.py (hunk 5 above)
Comment (laanwj) [reply test/functional/test_framework/socks5.py]: thanks!
factoring it out was more fo a readablity thing, i don't really mind if it's in netutil
i'm fine with leaving it where it is, if some other code needs it it can always be moved to a more general place

**Comment by jonatack** [comment]
Approach ACK 57529ac4dbb2721c1ad0a3566f0299dbdb5ca5c0

**Comment by achow101** [comment]
ACK 57529ac4dbb2721c1ad0a3566f0299dbdb5ca5c0

File: test/functional/test_framework/socks5.py [hunk 8]
```diff
@@ -117,6 +171,22 @@ def handle(self):
...
+                else:
+                    logger.debug(f"Closing connection to {requested_to}: the destinations factory returned None")
```
[the remark this answers is not in the document]
Comment (vasild) [reply test/functional/test_framework/socks5.py:186]: I think this is worth a followup. If you agree, I would open one, or just leave it as it is? What to reword it to? "Ending handling of connection to ..."? Or log different messages based on `self.serv.keep_alive`?

[the remark this answers is not in the document]
Comment (mzumsande) [reply test/functional/test_framework/socks5.py:186]: yes, happy to ack that follow-up. I think everything is fine, but my suggestion would be something like "Can't serve the connection to..." because we already log "Serving connection to..." in the success case. And maybe add another log entry "Keeping connection alive" `else` clause in the `finally` block a few lines below to avoid conditional log messages.

[the remark this answers is not in the document]
Comment (vasild) [reply test/functional/test_framework/socks5.py:186]: Done in https://github.com/bitcoin/bitcoin/pull/31239
